src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    return ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()));
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
+    return m_log_events.size();
+}
+
+auto ClpIrV1Decoder::build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    constexpr size_t cFullRangeEndIdx{0};
+    if (0 != begin_idx && cFullRangeEndIdx != end_idx) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Partial range index building is not yet supported."
+        );
+    }
+    if (false == m_full_range_built) {
+        m_full_range_built = true;
+        constexpr size_t cDefaultNumLogEvents{500'000};
+        m_log_events.reserve(cDefaultNumLogEvents);
+        while (true) {
+            auto const result{m_deserializer.deserialize_log_event()};
+            if (false == result.has_error()) {
+                m_log_events.emplace_back(result.value());
+                continue;
+            }
+            auto const error{result.error()};
+            if (std::errc::no_message_available == error) {
+                break;
+            }
+            if (std::errc::result_out_of_range == error) {
+                SPDLOG_ERROR("File contains an incomplete IR stream");
+                break;
+            }
+            throw ClpJsException(
+                    clp::ErrorCode::ErrorCode_Corrupt,
+                    __FILENAME__,
+                    __LINE__,
+                    "Failed to decompress: "s + error.category().name() + ":" + error.message()
+            );
+        }
+        m_data_buffer.reset(nullptr);
```

Comment:
This is technically a bit unsafe since the Zstandard decompressor is still open. A similar issue exists with passing the buffer and the Zstandard decompressor to the constructor separately. Since their life cycles are inherently linked, how about creating a small class to encapsulate the two together?

Reply:
Makes sense. I'll make the change once we reach a conclusion for https://github.com/y-scope/clp-ffi-js/pull/1/files/7939fd36ddc9a06a1e5e57ef2b27c2db774b1a71#r1704920663 .

Reply:
(Please correct me if I'm wrong with anything below.)

Now I think about it again, we probably also shouldn't only free the data buffer and the decompressor while leaving the deserializer opened, or the deserializer would hold a dangling reference to the reader (decompressor).

i.e., the lifecycles of the data buffer, the decompressor, and the deserializer are all inherently linked. If we're to create a class to encapsulate all, that class would probably look like the current StreamReader with the decoding stuff stripped.

On the other hand, we already have a `m_read_complete` boolean flag which prevents any unexpected read-after-free accesses. For now, do you think we can keep everything as-is and add the missing `m_zstd_decompressor->close();` after we free the `m_data_buffer`?

Reply:
> we probably also shouldn't only free the data buffer and the decompressor while leaving the deserializer opened, or the deserializer would hold a dangling reference to the reader (decompressor)

Right, I missed the link there too.

> If we're to create a class to encapsulate all, that class would probably look like the current StreamReader with the decoding stuff stripped.

Uh, I don't think we have to create a complicated class; my intention was just that the class encapsulates the closing logic together. I think one way to write this logic is to:

* Create a class just to store the buffer, Zstd compressor, and deserializer.
* The class should also provide a method to get a reference to the deserializer that we can use wherever necessary.
* We can store an instance of this class as a unique pointer in the stream reader.
* When the deserializer is exhausted, we can free the unique pointer.
* We can replace m_read_complete with checking the free-ness of the unique_pointer.

What do you think?